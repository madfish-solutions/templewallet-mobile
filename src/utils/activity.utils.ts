import { ActivityGroup, ActivityInterface, emptyActivity } from '../interfaces/activity.interface';

export const groupActivitiesByHash = (operations: ActivityInterface[], transfers: ActivityInterface[]) => {
  const result: ActivityGroup[] = [];

  const activities = operations.concat(transfers).sort((a, b) => b.timestamp - a.timestamp);
  let prevActivity: ActivityInterface = emptyActivity;

  for (const activity of activities) {
    if (activity.hash === prevActivity.hash) {
      result[result.length - 1].push(activity);
    } else {
      result.push([activity]);
    }

    prevActivity = activity;
  }

  return result;
};
