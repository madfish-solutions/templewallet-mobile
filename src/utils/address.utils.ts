export const truncateLongAddress = (data: string) =>
  data.length > 10 ? data.slice(0, 5) + '...' + data.slice(-5) : data;
