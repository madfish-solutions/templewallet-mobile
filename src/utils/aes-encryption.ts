/* eslint-disable @typescript-eslint/no-explicit-any */

// import * as crypto from 'crypto';
import { NativeModules } from 'react-native';

import { isAndroid } from 'src/config/system';

const AesEncryption = NativeModules.Aes;

export const encrypt = async (password: string, text: string) => {
  try {
    const salt = generateSalt(16);
    const key = await keyFromPassword(password, salt);
    const result = await encryptWithKey(text, key);
    result.salt = salt;

    return JSON.stringify(result);
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const decrypt = async (password: string, encryptedString: string) => {
  try {
    const encryptedData = JSON.parse(encryptedString);
    const key = await keyFromPassword(password, encryptedData.salt);
    const data = await decryptWithKey(encryptedData, key);

    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};

const generateSalt = (byteCount = 32) => {
  const view = new Uint8Array(byteCount);

  crypto.getRandomValues(view);

  const b64encoded = btoa(
    String.fromCharCode.apply(
      null,
      // @ts-expect-error
      view
    )
  );

  return b64encoded;
};

const generateKey = (password: string, salt: string) =>
  isAndroid ? AesEncryption.pbkdf2(password, salt, 5000, 256) : AesEncryption.pbkdf2(password, salt);

const keyFromPassword = (password: string, salt: string) => generateKey(password, salt);

const encryptWithKey = (text: string, keyBase64: any) => {
  const ivBase64 = generateSalt(32);

  return AesEncryption.encrypt(text, keyBase64, ivBase64).then((cipher: any) => ({
    cipher,
    iv: ivBase64
  }));
};

const decryptWithKey = (encryptedData: any, key: any) =>
  AesEncryption.decrypt(encryptedData.cipher, key, encryptedData.iv);

const btoa = (data: string) => Buffer.from(data).toString('base64');
