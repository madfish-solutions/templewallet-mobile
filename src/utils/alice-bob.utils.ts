import { templeWalletApi } from 'src/api.service';

export const getTezUahPairInfo = async () => {
  const { data } = await templeWalletApi.get<{ minAmount: number; maxAmount: number }>('/alice-bob/get-pair-info');

  return data;
};

export const getTezUahPairEstimation = async (uahAmount: number) => {
  const { data } = await templeWalletApi.post<{ outputAmount: number }>('/alice-bob/estimate-amount', {
    params: {
      amount: uahAmount
    }
  });

  return data.outputAmount;
};
