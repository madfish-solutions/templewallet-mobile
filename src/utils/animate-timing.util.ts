import { Animated, Easing } from 'react-native';

import { ANIMATION_DURATION } from '../config/animation';
import { emptyFn } from '../config/general';
import { Optional } from '../interfaces/optional.type';

export type AnimateTimingConfig = Optional<Animated.TimingAnimationConfig, 'useNativeDriver'>;

export const animateTiming = (
  animation: Animated.AnimatedValue | Animated.AnimatedValueXY,
  {
    toValue,
    easing = Easing.linear,
    duration = ANIMATION_DURATION,
    delay = 0,
    useNativeDriver = true
  }: AnimateTimingConfig,
  callback = emptyFn
) =>
  Animated.timing(animation, {
    toValue,
    duration,
    easing,
    delay,
    useNativeDriver
  }).start(callback);
