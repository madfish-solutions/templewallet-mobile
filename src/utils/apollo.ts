import { ApolloClient, InMemoryCache, HttpLink, FetchResult, OperationVariables, QueryOptions } from '@apollo/client';
import fetch from 'cross-fetch';
import { DocumentNode } from 'graphql';
import { from } from 'rxjs';
import { filter } from 'rxjs/operators';

import { isDefined } from './is-defined';

export const buildApolloClient = (uri: string) =>
  new TempleApolloClient({
    link: new HttpLink({ uri, fetch }),
    cache: new InMemoryCache()
  });

class TempleApolloClient<TCacheShape> extends ApolloClient<TCacheShape> {
  async fetch<T, TVars = OperationVariables>(query: DocumentNode, variables?: TVars, options?: QueryOptions<TVars, T>) {
    const result: FetchResult<T> = await super.query<T, TVars>({
      query,
      variables,
      // Disabling cache as it creates bottlenecks (blocks thread) when fetching large data
      fetchPolicy: 'no-cache',
      ...options
    });

    return result.data;
  }

  fetch$<T, TVars = OperationVariables>(query: DocumentNode, variables?: TVars, options?: QueryOptions<TVars, T>) {
    return from(this.fetch<T, TVars>(query, variables, options)).pipe(filter(isDefined));
  }
}
