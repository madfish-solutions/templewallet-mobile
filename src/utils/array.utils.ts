export const sliceIntoChunks = <T>(array: T[], chunkSize: number) => {
  const result = [];

  for (let i = 0; i < array.length; i += chunkSize) {
    const chunk = array.slice(i, i + chunkSize);
    result.push(chunk);
  }

  return result;
};
