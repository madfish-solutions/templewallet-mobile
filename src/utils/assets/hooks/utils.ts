import { VisibilityEnum } from 'src/enums/visibility.enum';
import { AccountTokenInterface } from 'src/token/interfaces/account-token.interface';
import { TokenMetadataInterface } from 'src/token/interfaces/token-metadata.interface';

import { UsableAccountAsset } from '../types';

export const buildUsableAccountAsset = (
  asset: AccountTokenInterface,
  metadata: TokenMetadataInterface,
  exchangeRate?: number
): UsableAccountAsset => {
  let visibility = asset.visibility;
  if (visibility === VisibilityEnum.InitiallyHidden && Number(asset.balance) > 0) {
    visibility = VisibilityEnum.Visible;
  }

  return {
    slug: asset.slug,
    visibility,
    balance: asset.balance,
    exchangeRate,
    ...metadata
  };
};
