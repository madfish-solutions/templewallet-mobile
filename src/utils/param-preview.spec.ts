import {
  mockContractCallParams,
  mockDelegationParams,
  mockFA1_2ApproveParams,
  mockFA1_2TokenTransferParams,
  mockFA_2TokenTransferParams,
  mockOriginationParams,
  mockTezosTransferParams
} from '../interfaces/op-params.interface.mock';

import { getParamPreview } from './param-preview.utils';

describe('getParamPreview', () => {
  it('should parse delegation operation', () => {
    expect(getParamPreview(mockDelegationParams)).toMatchSnapshot();
  });

  it('should parse transfer of TEZ', () => {
    expect(getParamPreview(mockTezosTransferParams)).toMatchSnapshot();
  });

  it('should parse transfer of FA1.2 token', () => {
    expect(getParamPreview(mockFA1_2TokenTransferParams)).toMatchSnapshot();
  });

  it('should parse transfer of FA2 token', () => {
    expect(getParamPreview(mockFA_2TokenTransferParams)).toMatchSnapshot();
  });

  it('should parse approval of FA1.2 token expense', () => {
    expect(getParamPreview(mockFA1_2ApproveParams)).toMatchSnapshot();
  });

  it('should provide a fallback for contract calls', () => {
    expect(getParamPreview(mockContractCallParams)).toMatchSnapshot();
  });

  it('should provide a fallback for other operations', () => {
    expect(getParamPreview(mockOriginationParams)).toMatchSnapshot();
  });
});
