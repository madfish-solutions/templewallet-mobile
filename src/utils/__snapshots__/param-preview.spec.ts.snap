// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParamPreview should parse approval of FA1.2 token expense 1`] = `
Object {
  "amount": "20",
  "approveTo": "contractAddress1",
  "asset": Object {
    "contract": "fa12TokenAddress",
  },
  "type": "FA1_2Approve",
}
`;

exports[`getParamPreview should parse delegation operation 1`] = `
Object {
  "baker": "delegate1",
  "type": "Delegate",
}
`;

exports[`getParamPreview should parse transfer of FA1.2 token 1`] = `
Object {
  "transfers": Array [
    Object {
      "amount": "20",
      "asset": Object {
        "contract": "fa12TokenAddress",
      },
      "recipient": "address2",
    },
  ],
  "type": "Send",
}
`;

exports[`getParamPreview should parse transfer of FA2 token 1`] = `
Object {
  "transfers": Array [
    Object {
      "amount": "1000",
      "asset": Object {
        "contract": "fa2TokenAddress",
        "id": 10,
      },
      "recipient": "address2",
    },
  ],
  "type": "Send",
}
`;

exports[`getParamPreview should parse transfer of TEZ 1`] = `
Object {
  "transfers": Array [
    Object {
      "amount": "10",
      "asset": "tez",
      "recipient": "address2",
    },
  ],
  "type": "Send",
}
`;

exports[`getParamPreview should provide a fallback for contract calls 1`] = `
Object {
  "amount": "0",
  "contract": "contractAddress1",
  "entrypoint": "foo",
  "type": "ContractCall",
}
`;

exports[`getParamPreview should provide a fallback for other operations 1`] = `
Object {
  "opKind": "origination",
  "type": "Other",
}
`;
